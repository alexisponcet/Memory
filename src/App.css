.memory {
    margin: auto;
    padding: 0;
    width: 400px;
    height: 500px;

    display: flex;
    flex-wrap: wrap;
    user-select: none;
}