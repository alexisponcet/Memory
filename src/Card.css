.memory > .card {
    margin: 0.2em;
    padding: 0;

    font-size: 2em;
    display: flex;
    flex: 1 1 calc(100% / 8 - 0.4em);
    outline: 0.1em solid silver;
    cursor: default;
}

.memory > .card.hidden {
    background: black;
}

.memory > .card.justMatched {
    outline-color: green;
}
.memory > .card.justMismatched {
    outline-color: red;
}

.memory > .card.visible {
    cursor: not-allowed;
}

.memory > .card > .symbol {
    margin: auto;
}