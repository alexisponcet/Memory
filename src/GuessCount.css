.memory > .guesses {
    margin: 0.2em;
    padding: 0;

    width: calc(100% - 0.4em);

    font-size: 2em;
    font-family: monospace;
    text-align: right;
}