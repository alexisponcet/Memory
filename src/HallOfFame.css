.hallOfFame {
    width: calc(100% - 0.4em);;
    border-collapse: collapse;
}

.hallOfFame .date,
.hallOfFame .guesses {
    width: 20%;
    text-align: center;
}

.hallOfFame tr:nth-child(even) {
    background: #eee;
}