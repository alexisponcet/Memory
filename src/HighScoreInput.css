.highScoreInput {
    margin: 0.2em;

    width: calc(100% - 0.4em);

    font-size: 1em;
}

.highScoreInput input {
    margin-bottom: 0.2em;

    width: calc(100% - 0.4em);
    display: block;

    font-family: monospace;
}