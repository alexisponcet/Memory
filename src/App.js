import React, { Component } from 'react';
import shuffle from 'lodash.shuffle';

import './App.css';

import Card from './Card';
import GuessCount from './GuessCount';
import HallOfFame from './HallOfFame';
import HighScoreInput from './HighScoreInput';

export const SIDE = 8;
export const SYMBOLS = '🍞🥐🥖🥞🧀🍖🍗🥓🍔🍟🍕🌭🌮🌯🥙🥚🍳🥘🍲🥗🍿🍇🍈🍉🍊🍋🍌🍍🍎🍏🍐🍑';
export const VISUAL_PAUSE_MSECS = 750;

class App extends Component {

	state = {
		cards: this.generateCards(),
		currentPair: [],
		guesses: 0,
		hallOfFame: null,
		matchedCardIndices: [],
	}

	/**
	 * Initialize the array of hidden cards
	 * @returns Array : the array of hidden cards (shuffled)
	 */
    generateCards() {
        const result = [];
        const size = SIDE * SIDE;
        const candidates = shuffle(SYMBOLS);
        while (result.length < size) {
            const card = candidates.pop();
            result.push(card, card);
        }
        return shuffle(result);
    }

	/**
	 * Check if the player has chosen 2 cards
	 * @param index : the index of the last clicked card
	 */
    handleCardClick = index => {
        const { currentPair } = this.state;

        if (currentPair.length === 2) {
            return;
        }

        if (currentPair.length === 0) {
            this.setState({ currentPair: [index] });
            return;
        }

        this.handleNewPairClosedBy(index);
    }

	/**
	 * Check if the player has found a pair of cards
	 * @param index : the index of the last clicked card
	 */
	handleNewPairClosedBy(index){
		const { cards, currentPair, guesses, matchedCardIndices } = this.state;

		const newGuesses = guesses + 1;
		const newPair = [currentPair[0], index];
		const matched = (cards[newPair[0]] === cards[newPair[1]]
							&& newPair[0] !== newPair[1]);
		this.setState({guesses : newGuesses, currentPair : newPair});
		if (matched){
			this.setState({matchedCardIndices: [...matchedCardIndices, ...newPair]});
		}
		setTimeout(() => this.setState({ currentPair: [] }), VISUAL_PAUSE_MSECS);
	}

	/**
	 * Check the state of the card
	 * @param index : the index of the card
	 * @returns String : the state of the card
	 */
	getFeedbackForCard(index) {
        const { currentPair, matchedCardIndices } = this.state;
        const indexMatched = matchedCardIndices.includes(index);

        if (currentPair.length < 2) {
            return indexMatched || index === currentPair[0] ? 'visible' : 'hidden';
        }

        if (currentPair.includes(index)) {
            return indexMatched ? 'justMatched' : 'justMismatched';
        }

        return indexMatched ? 'visible' : 'hidden';
    }

	/**
	 * Display the hall of fame
	 * @param hallOfFame : the hall of fame after being updated
	 */
	displayHallOfFame = (hallOfFame) => {
		this.setState({ hallOfFame });
	}

    render() {
        const { cards, guesses, hallOfFame, matchedCardIndices } = this.state;
        const won = matchedCardIndices.length === cards.length;
        return (
            <div className="memory">
                <GuessCount guesses = {guesses} />
                {cards.map((card, index) =>
                    <Card card={card}
                          feedback={this.getFeedbackForCard(index)}
                          index={index}
                          key={index}
                          onClick={this.handleCardClick}
                    />
                )}
	            {
		            won &&
		            (hallOfFame ? (
			            <HallOfFame entries={hallOfFame} />
		            ) : (
			            <HighScoreInput guesses={guesses} onStored={this.displayHallOfFame} />
		            ))
	            }
            </div>
        )
    }
}

export default App;