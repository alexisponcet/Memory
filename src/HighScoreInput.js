import React, { Component } from 'react';

import './HighScoreInput.css';

import PropTypes from 'prop-types';
import { saveHOFEntry } from './HallOfFame';

class HighScoreInput extends Component {

	state = {
		winner : '',
	}

	/**
	 * Update the local state with the name of the current player
	 * @param value : the name of the player
	 */
	handleWinnerUpdate = ( {target : {value}}) => {
		this.setState({ winner: value.toUpperCase() });
	}

	/**
	 * Build the new player and store it in the hall of fame
	 * @param event : from the form (disable the default behaviour)
	 */
	persistWinner = (event) => {
		event.preventDefault();
		const newEntry = { guesses: this.props.guesses, player: this.state.winner };
		saveHOFEntry(newEntry, this.props.onStored);
	}

	render() {
		return (
			<form className="highScoreInput" onSubmit={this.persistWinner}>
				<p>
					<label>
						Congratulatation ! Type your username :
						<input type="text"
						       autoComplete="given-name"
						       value = {this.state.winner}
							   onChange={this.handleWinnerUpdate}/>
					</label>
					<button type="submit">Let's go !</button>
				</p>
			</form>
		)
	}
}

HighScoreInput.propTypes = {
	guesses: PropTypes.number.isRequired,
	onStored: PropTypes.func.isRequired,
}

export default HighScoreInput;