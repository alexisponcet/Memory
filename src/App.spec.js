import React from 'react';
import { expect } from 'chai';
import { shallow } from 'enzyme';
import sinon from 'sinon';

import App, {SIDE, SYMBOLS, VISUAL_PAUSE_MSECS} from './App';
import GuessCount from './GuessCount';
import Card from './Card';
import HallOfFame from './HallOfFame';
import HighScoreInput from './HighScoreInput';

describe('<App />', () => {

	describe('Structure App', () => {
		// Initial State : won === false
		describe('when starting a game', () => {
			const wrapper = shallow(<App/>);

			it('has a guessCount with 0', () => {
				expect(wrapper).to.have.descendants(GuessCount);

				expect(wrapper.find(GuessCount)).to.have.prop('guesses').deep.equal(0);
			})

			it('has 36 cards', () => {
				expect(wrapper.find(Card)).to.have.lengthOf(SIDE * SIDE);
			})
		})

		// Final State : won === true
		describe('when finishing a game', () => {

			describe('when first one', () => {
				const wrapper = shallow(<App/>);
				const guesses = 20;
				wrapper.setState({matchedCardIndices : [], cards : [], hallOfFame : null, guesses : 20});

				it('has new HighScoreInput', () => {
					// callback
					expect(wrapper).to.have.descendants(HighScoreInput);
					expect(wrapper.find(HighScoreInput)).to.have.prop('guesses').deep.equal(guesses);
				})
			})

			describe('when not first one', () => {
				const wrapper = shallow(<App/>);
				const hallOfFame = [{id: 0, guesses: 22, date: '08/09/2018', player: 'Jane' }];
				wrapper.setState({matchedCardIndices : [], cards : [], hallOfFame : hallOfFame});

				it('has Hall of Fame', () => {
					expect(wrapper).to.have.descendants(HallOfFame);
					expect(wrapper.find(HallOfFame)).to.have.prop('entries').deep.equal(hallOfFame);
				})
			})
		})
	})

	describe('when a card is clicked', () => {
		const stub = sinon.stub(App.prototype, 'generateCards')
			.returns([...SYMBOLS.repeat(2)]);
		const wrapper = shallow(<App/>);
		const instance = wrapper.instance();
		const firstCard = 0, secondCard = 1, thirdCard = 2;
		const firstCardMatched = (SIDE*SIDE)/2 + firstCard;

		it('add the first card in the current list of cards', () => {
			wrapper.setState({guesses : 0, currentPair: []});

			instance.handleCardClick(firstCard);

			expect(JSON.stringify(wrapper.state().currentPair)).to.equal(JSON.stringify([firstCard]));
		})

		it('add the second card in the current list of cards and ' +
			'do not match with the first one', () => {
			wrapper.setState({guesses : 0, currentPair: [firstCard]});

			instance.handleCardClick(secondCard);

			expect(JSON.stringify(wrapper.state().currentPair)).to.equal(JSON.stringify([firstCard, secondCard]));
			expect(wrapper.state().guesses).to.equal(1);
		})

		it('add the second card in the current list of cards and ' +
			'match with the first one', () => {
			wrapper.setState({guesses : 0, currentPair: [firstCard]});

			instance.handleCardClick(firstCardMatched);

			expect(JSON.stringify(wrapper.state().matchedCardIndices)).to.equal(JSON.stringify([firstCard, firstCardMatched]));
			expect(wrapper.state().guesses).to.equal(1);
		})

		it('do not add the third card in the current list of card', () => {
			wrapper.setState({guesses : 0, currentPair: [firstCard, secondCard]});

			instance.handleCardClick(thirdCard);

			expect(wrapper.state().currentPair).to.length(2);
		})

		it('reset current list of cards after 750ms', () => {
			const clock = sinon.useFakeTimers();
			const wrapper = shallow(<App/>);

			wrapper.instance().handleNewPairClosedBy(thirdCard);
			clock.tick(VISUAL_PAUSE_MSECS);

			expect(wrapper.state().currentPair).to.length(0);
			clock.restore();
		})

		it('should match its reference snapshot', () => {
			expect(wrapper).to.matchSnapshot();
			stub.restore();
		})
	})
})