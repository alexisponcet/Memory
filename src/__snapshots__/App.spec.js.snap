// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> when a card is clicked should match its reference snapshot 1`] = `
<div
  className="memory"
>
  <GuessCount
    guesses={0}
  />
  <Card
    card="🍞"
    feedback="justMatched"
    index={0}
    key="0"
    onClick={[Function]}
  />
  <Card
    card="🥐"
    feedback="justMismatched"
    index={1}
    key="1"
    onClick={[Function]}
  />
  <Card
    card="🥖"
    feedback="hidden"
    index={2}
    key="2"
    onClick={[Function]}
  />
  <Card
    card="🥞"
    feedback="hidden"
    index={3}
    key="3"
    onClick={[Function]}
  />
  <Card
    card="🧀"
    feedback="hidden"
    index={4}
    key="4"
    onClick={[Function]}
  />
  <Card
    card="🍖"
    feedback="hidden"
    index={5}
    key="5"
    onClick={[Function]}
  />
  <Card
    card="🍗"
    feedback="hidden"
    index={6}
    key="6"
    onClick={[Function]}
  />
  <Card
    card="🥓"
    feedback="hidden"
    index={7}
    key="7"
    onClick={[Function]}
  />
  <Card
    card="🍔"
    feedback="hidden"
    index={8}
    key="8"
    onClick={[Function]}
  />
  <Card
    card="🍟"
    feedback="hidden"
    index={9}
    key="9"
    onClick={[Function]}
  />
  <Card
    card="🍕"
    feedback="hidden"
    index={10}
    key="10"
    onClick={[Function]}
  />
  <Card
    card="🌭"
    feedback="hidden"
    index={11}
    key="11"
    onClick={[Function]}
  />
  <Card
    card="🌮"
    feedback="hidden"
    index={12}
    key="12"
    onClick={[Function]}
  />
  <Card
    card="🌯"
    feedback="hidden"
    index={13}
    key="13"
    onClick={[Function]}
  />
  <Card
    card="🥙"
    feedback="hidden"
    index={14}
    key="14"
    onClick={[Function]}
  />
  <Card
    card="🥚"
    feedback="hidden"
    index={15}
    key="15"
    onClick={[Function]}
  />
  <Card
    card="🍳"
    feedback="hidden"
    index={16}
    key="16"
    onClick={[Function]}
  />
  <Card
    card="🥘"
    feedback="hidden"
    index={17}
    key="17"
    onClick={[Function]}
  />
  <Card
    card="🍲"
    feedback="hidden"
    index={18}
    key="18"
    onClick={[Function]}
  />
  <Card
    card="🥗"
    feedback="hidden"
    index={19}
    key="19"
    onClick={[Function]}
  />
  <Card
    card="🍿"
    feedback="hidden"
    index={20}
    key="20"
    onClick={[Function]}
  />
  <Card
    card="🍇"
    feedback="hidden"
    index={21}
    key="21"
    onClick={[Function]}
  />
  <Card
    card="🍈"
    feedback="hidden"
    index={22}
    key="22"
    onClick={[Function]}
  />
  <Card
    card="🍉"
    feedback="hidden"
    index={23}
    key="23"
    onClick={[Function]}
  />
  <Card
    card="🍊"
    feedback="hidden"
    index={24}
    key="24"
    onClick={[Function]}
  />
  <Card
    card="🍋"
    feedback="hidden"
    index={25}
    key="25"
    onClick={[Function]}
  />
  <Card
    card="🍌"
    feedback="hidden"
    index={26}
    key="26"
    onClick={[Function]}
  />
  <Card
    card="🍍"
    feedback="hidden"
    index={27}
    key="27"
    onClick={[Function]}
  />
  <Card
    card="🍎"
    feedback="hidden"
    index={28}
    key="28"
    onClick={[Function]}
  />
  <Card
    card="🍏"
    feedback="hidden"
    index={29}
    key="29"
    onClick={[Function]}
  />
  <Card
    card="🍐"
    feedback="hidden"
    index={30}
    key="30"
    onClick={[Function]}
  />
  <Card
    card="🍑"
    feedback="hidden"
    index={31}
    key="31"
    onClick={[Function]}
  />
  <Card
    card="🍞"
    feedback="visible"
    index={32}
    key="32"
    onClick={[Function]}
  />
  <Card
    card="🥐"
    feedback="hidden"
    index={33}
    key="33"
    onClick={[Function]}
  />
  <Card
    card="🥖"
    feedback="hidden"
    index={34}
    key="34"
    onClick={[Function]}
  />
  <Card
    card="🥞"
    feedback="hidden"
    index={35}
    key="35"
    onClick={[Function]}
  />
  <Card
    card="🧀"
    feedback="hidden"
    index={36}
    key="36"
    onClick={[Function]}
  />
  <Card
    card="🍖"
    feedback="hidden"
    index={37}
    key="37"
    onClick={[Function]}
  />
  <Card
    card="🍗"
    feedback="hidden"
    index={38}
    key="38"
    onClick={[Function]}
  />
  <Card
    card="🥓"
    feedback="hidden"
    index={39}
    key="39"
    onClick={[Function]}
  />
  <Card
    card="🍔"
    feedback="hidden"
    index={40}
    key="40"
    onClick={[Function]}
  />
  <Card
    card="🍟"
    feedback="hidden"
    index={41}
    key="41"
    onClick={[Function]}
  />
  <Card
    card="🍕"
    feedback="hidden"
    index={42}
    key="42"
    onClick={[Function]}
  />
  <Card
    card="🌭"
    feedback="hidden"
    index={43}
    key="43"
    onClick={[Function]}
  />
  <Card
    card="🌮"
    feedback="hidden"
    index={44}
    key="44"
    onClick={[Function]}
  />
  <Card
    card="🌯"
    feedback="hidden"
    index={45}
    key="45"
    onClick={[Function]}
  />
  <Card
    card="🥙"
    feedback="hidden"
    index={46}
    key="46"
    onClick={[Function]}
  />
  <Card
    card="🥚"
    feedback="hidden"
    index={47}
    key="47"
    onClick={[Function]}
  />
  <Card
    card="🍳"
    feedback="hidden"
    index={48}
    key="48"
    onClick={[Function]}
  />
  <Card
    card="🥘"
    feedback="hidden"
    index={49}
    key="49"
    onClick={[Function]}
  />
  <Card
    card="🍲"
    feedback="hidden"
    index={50}
    key="50"
    onClick={[Function]}
  />
  <Card
    card="🥗"
    feedback="hidden"
    index={51}
    key="51"
    onClick={[Function]}
  />
  <Card
    card="🍿"
    feedback="hidden"
    index={52}
    key="52"
    onClick={[Function]}
  />
  <Card
    card="🍇"
    feedback="hidden"
    index={53}
    key="53"
    onClick={[Function]}
  />
  <Card
    card="🍈"
    feedback="hidden"
    index={54}
    key="54"
    onClick={[Function]}
  />
  <Card
    card="🍉"
    feedback="hidden"
    index={55}
    key="55"
    onClick={[Function]}
  />
  <Card
    card="🍊"
    feedback="hidden"
    index={56}
    key="56"
    onClick={[Function]}
  />
  <Card
    card="🍋"
    feedback="hidden"
    index={57}
    key="57"
    onClick={[Function]}
  />
  <Card
    card="🍌"
    feedback="hidden"
    index={58}
    key="58"
    onClick={[Function]}
  />
  <Card
    card="🍍"
    feedback="hidden"
    index={59}
    key="59"
    onClick={[Function]}
  />
  <Card
    card="🍎"
    feedback="hidden"
    index={60}
    key="60"
    onClick={[Function]}
  />
  <Card
    card="🍏"
    feedback="hidden"
    index={61}
    key="61"
    onClick={[Function]}
  />
  <Card
    card="🍐"
    feedback="hidden"
    index={62}
    key="62"
    onClick={[Function]}
  />
  <Card
    card="🍑"
    feedback="hidden"
    index={63}
    key="63"
    onClick={[Function]}
  />
</div>
`;
